cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogFileManager.java

```diff
@@ -165,8 +168,16 @@ public void close() {
   @Override
   public void flush() throws IOException {
     // perform flush on all the files in the outputStreamMap
-    for (LogFileOutputStream file : outputStreamMap.values()) {
-      file.flush();
+    long currentTs = System.currentTimeMillis();
+    List<LogFileOutputStream> outputStreamList = new ArrayList<>(outputStreamMap.values());
```

Comment:
This method is different then close. Flush gets called much more frequent, and copying like this is not free. What you can do is to use the iterator:

```java
Iterator<LogFileOutputStream> itor = outputStreamMap.values().iterator();
while (itor.hasNext()) {
  LogFileOutputStream stream = itor.next();
  stream.flush();
  long timeSinceFileCreated = currentTs - stream.getCreateTime();
  if (timeSinceFileCreated > maxLifetimeMillis) {
    // Remove it from the map first. This also make sure even if the close failed, the stream won't stays in the map
    itor.remove();
    logFileOutputStream.close();
  }
}